### ISSUE TITLE
[Feature]: Autoload TimeSeries data with offset, conversion and channel conversion

### ISSUE THREAD

**Issue by nmarkowitz** [NeurodataWithoutBorders/pynwb#1474]
### What would you like to see added to PyNWB?

Hi,

As pynwb gets has more properties to `TimeSeries` data added it would be convenient if the load function automatically applies the the conversion if required. 

### What solution would you like?

For a typical `TimeSeries` this could look something like

```
from pynwb import NWBHDF5IO
import numpy as np

io = NWBHDF5IO('myfile.nwb',mode='r',load_namespaces=True)
nwbfile = io.read()

% Get TimeSeries
ts = nwbfile.acquisition['test_ts']
ts_keys = test_ts.fields.keys()
data = ts.data[()]
nchans = data.shape[1]

% Set default offset, conversion and channel_conversion
conversion = ts.conversion if 'convervsion' in ts_keys else 1
offset = ts.offset if 'convervsion' in ts_keys else 0
channel_conversion = ts.channel_conversion if 'channel_conversion' in ts_keys else np.ones((1,nchans))

% Run conversion
data = data*channel_conversion*conversion + offset
```

This example isn't perfect but it would make it backwards compatible and does not multiline `if` statements. It would probably also be good to still have the option of loading the data without running conversion. Whether that would be a hidden function (such as `ts._load()`) or adding an extra argument to the load function I'm not sure.

### Do you have any interest in helping implement the feature?

Yes, but I would need guidance.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/.github/CODE_OF_CONDUCT.rst)
- [X] Have you checked the [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [X] Have you ensured this change was not already [requested](https://github.com/NeurodataWithoutBorders/pynwb/issues)?

**Comment by oruebel** [comment]
I would need to take a closer look, but I don't think ``TimeSeries`` has a ``__getitem__`` function defined right now. If so, then an alternate solution would be to implement the conversion as part of ``TimeSeries.__getitem__`` such that slicing with ``ts.data[0:10]`` would remain as is and return raw values and ``ts[0:10]`` would return converted values. Another interesting functionality this would allow for is to then also return converted timestamps along with the data when slicing with ``ts[0:10]`` (e.g., in the case that the TimeSeries has a ``starting_time`` and ``rate`` defined, the user similarly needs to compute the correct timestamps).

**Comment by nmarkowitz** [comment]
@oruebel implementing a `__getitem__` function sounds like a good idea. If you'll be loading timestamps as well then it would make sense also load the `control` values as well if they are defined (maybe any field that has a single value per sample).

For timestamps, that could work but is also conditional in that `rate` and not `timestamps` is used to define time. When `timestamps` are used I try to see if most of the timestamps are continuous with only a few that show gaps (ex: the median difference in time between timestamps is 0.0033 and therefore the data can be considered to have a rate 300Hz). I then add those timestamps that are missing and then add `NaN` values to the data in place where those synthetic timestamps were inserted. This is just a thought but it is something I have encountered and the method I've used to treat data as continuous for the many toolboxes that require continuous data.

**Comment by oruebel** [comment]
> . If you'll be loading timestamps as well then it would make sense also load the `control` values as well if they are defined

Good point. I think maybe the behavior of which fields to load may need to be configurable in some form for ``TimeIntervals.__getitem__`` (not sure yet).

> For timestamps, that could work but is also conditional in that `rate` and not `timestamps` is used to define time

My thinking here was that this is basically is the same issue as with the conversion of ``data``. I.e., ``TimeSeries.__getitem__`` would similarly always return full timestamps that match the selection, i.e., if ``starting_time`` and ``rate`` are used we would construct the timestamps as list or generator and otherwise we would load the data from ``timestamps``. 

> When `timestamps` are used I try to see if most of the timestamps are continuous with only a few that show gaps (ex: the median difference in time between timestamps is 0.0033 and therefore the data can be considered to have a rate 300Hz).

It seems like this is mainly relevant to how the file is being created/written rather than how we read the timeseries data. Just for reference, some related best practices for this are described in the NWBInspector here:

- https://nwbinspector.readthedocs.io/en/dev/best_practices/time_series.html#timestamps-vs-start-rate
- https://nwbinspector.readthedocs.io/en/dev/best_practices/time_series.html#timestamps-vs-start-rate

> I then add those timestamps that are missing and then add `NaN` values to the data in place where those synthetic timestamps were inserted.

If you are adding invalid data then you should also consider defining those in the ``InvalidTimesTable`` to make this behavior explicit:

- https://pynwb.readthedocs.io/en/stable/pynwb.file.html#pynwb.file.InvalidTimesTable
- https://nwb-schema.readthedocs.io/en/latest/format.html#groups-intervals-invalid-times

**Comment by nmarkowitz** [comment]
Thanks for sharing those links. I may not have known about the `InvalidTimesTable` object and wish I did sooner before converting some data. I'll refer to these links going forward. 

Going back to the `TimeIntervals.__getitem__` method one thought is that maybe it can just iterate over all fields in a `TimeSeries` and check if that fields contains a piece of data per sample to accomodate extensions. Or maybe define the method with some `kwargs` such as

```
def __getitem(self, data_index, return_timestamps=False, return_control=False, extra_fields=[]):
    % Get the data
```

Where `extra_fields` is an iterable containing the names of any other fields with a piece of data per sample point.

**Comment by oruebel** [comment]
> Or maybe define the method with some `kwargs` such as

I agree that some form of configurability will be required to define which fields to use and that extending the function to more specific TimeSeries types should be supported as well. In addition to slicing in time, the function will likely also need to support other fields.  E.g., for ``ElectricalSeries`` one would probably want to not just slice the time dimension but also the electrodes dimension.

**Comment by bendichter** [comment]
We've been maintaining some functions for common `TimeSeries` operations [here](https://github.com/NeurodataWithoutBorders/nwb-jupyter-widgets/blob/master/nwbwidgets/utils/timeseries.py#L7)

I think several of these could be refactored into methods of `TimeSeries` within pynwb:
* get_timeseries_tt
* get_timeseries_maxt
* get_timeseries_mint
* get_timeseries_in_units
* timeseries_time_to_ind

using `__getitem__` would certainly be convenient. I'm ambivalent as I worry a bit that it might be confusing to users who would expect `ts[:]` to do the same thing as `ts.data[:]`
